### PR TITLE
Add first iteration of ensemble evaluator

### LINKED ISSUE #1035: Propose first iteration of evaluation API

### PR DESCRIPTION
**Pull request by jondequinor**
Resolves #1035

### CODE REVIEW

File: ee.html [hunk 1]
```diff
@@ -0,0 +1,135 @@
+<!DOCTYPE html>
```
Comment (sondreso) [inline ee.html:1]: Where should we place this file?

Comment (jondequinor) [reply ee.html:1]: Could eventually be part of the docs as a sort of demo evaluator?

File: ert_shared/tracker/factory.py [hunk 2]
```diff
@@ -1,47 +1,59 @@
...
     do tracking.
+
+    If @ee_host_port_tuple then the factory will produce something that can
+    track an ensemble evaluator.
     """
     if num_realizations is not None:
         general_interval, detailed_interval = scale_intervals(num_realizations)
 
+    ee_monitor_connection_details = (
+        ("localhost", "8765")
```
Comment (markusdregi) [inline ert_shared/tracker/factory.py:35]: Magically appearing constants? :)

**Comment by markusdregi** [comment]
Note: This will not be pip-installable due to outdated libres. We should create a new tag in libres and get that tag into ERT as a lower bound for libres version...

File: ert_shared/ensemble_evaluator/evaluator.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        self._clients = set()
+        self._dispatchers_connected = asyncio.Queue(loop=self._loop)
+
+        self._snapshot = self.create_snapshot(ensemble)
+        self._event_index = 1
+
+    @staticmethod
+    def create_snapshot(ensemble):
+        builder = SnapshotBuilder()
+        real = ensemble.get_reals()[0]
```
Comment (sondreso) [inline ert_shared/ensemble_evaluator/evaluator.py:46]: This needs to be guarded

File: ee.html (hunk 1 above)
Comment (markusdregi) [reply ee.html:1]: Is it useful to have it around for you? If not, should we leave it in a separate branch for the moment?

File: ert_shared/ensemble_evaluator/config.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    "host": DEFAULT_HOST,
+    "port": DEFAULT_PORT,
+    "url": DEFAULT_URL,
+    "client_url": f"{DEFAULT_URL}/{CLIENT_URI}",
+    "dispatch_url": f"{DEFAULT_URL}/{DISPATCH_URI}",
+}
+
+
+def load_config(config_path=CONFIG_FILE):
+    if not os.path.exists(config_path):
```
Comment (markusdregi) [inline ert_shared/ensemble_evaluator/config.py:23]: Are we sure we want to be this forgiving? It seems like the code currently only uses the default, but there is no trace of an `ee_config.yml`. Hence, I guess `DEFAULT_EE_CONFIG` is used. Should we instead default `config_path=None` and then try to open `config_path` if it is not `None`? And fail if it does not exist?

**Review by markusdregi — COMMENTED** [review]
I've now made a go through of the PR and I think it looks very good! 👏 

One initial question (beyond the bike shedding while reading through it), have you tried to check that the test coverage of all exposed functionality is tested decently both with respect to happy and not so happy paths?

**Review by markusdregi — COMMENTED** [review]
Had two additional comments, but besides that I think this is looking good 👍

File: ert_gui/simulation/run_dialog.py [hunk 5]
```diff
@@ -207,8 +213,12 @@ def _on_tracker_event(self, event):
                     self.legends[state].updateLegend(state.name, 0, 0)
             else:
                 for state in event.sim_states:
-                    self.progress.updateState(
-                        state.state, 100.0 * state.count / state.total_count)
+                    try:
+                        self.progress.updateState(
+                            state.state, 100.0 * state.count / state.total_count)
+                    except ZeroDivisionError:
+                        # total_count not set by some slow tracker (EE)
```
[the remark this answers is not in the document]
Comment (jondequinor) [reply ert_gui/simulation/run_dialog.py:220]: It gets scarier further up the chain; If the EE doesn't produce events within 5 seconds of simulation start, the whole progress tracking thread dies.

We mitigate this in `legacy-EE` by sending ensemble level events in `evaluate`. That event should then mean that the `total_count` is set to `0`. 

So this `except` should go away.

[the remark this answers is not in the document]
Comment (markusdregi) [reply ert_gui/simulation/run_dialog.py:220]: Aight, so this goes entirely away when you merge in the legacy-EE, which is still the plan before we merge this into master?

File: ert_shared/ensemble_evaluator/evaluator.py [hunk 6]
```diff
@@ -21,6 +21,8 @@
 
 
 class EnsembleEvaluator:
+    dispatch = Dispatcher()
```
[the remark this answers is not in the document]
Comment (sondreso) [reply ert_shared/ensemble_evaluator/evaluator.py:24]: No, I will make it private!

File: ert_gui/simulation/run_dialog.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jondequinor) [reply ert_gui/simulation/run_dialog.py:220]: That's the plan!

**Comment by markusdregi** [comment]
I'm willing to approve this PR now. Do you still want to merge the `legacy` branch first? And in addition, feel free to rebase history at will...

**Comment by sondreso** [comment]
Yes, I would like to merge that first and then rebase, you want to take a look at that PR now and then I can rebase and squash everything afterwards? https://github.com/equinor/ert/pull/1148#pullrequestreview-539551329

**Comment by markusdregi** [comment]
Sounds like a plan 👍

**Review by markusdregi — APPROVED** [review]
Really good work 🎖️

**Pull request merged**